File: src/cloudtwin/providers/aws/sns/service.py
"""
SNS domain service.

Business logic for SNS operations. Has no knowledge of HTTP or XML.
Depends on repository interfaces only.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from cloudtwin.core.errors import NotFoundError
from cloudtwin.core.telemetry import TelemetryEngine
from cloudtwin.persistence.models import SnsMessage, SnsSubscription, SnsTopic
from cloudtwin.persistence.repositories import (
    SnsMessageRepository,
    SnsSubscriptionRepository,
    SnsTopicRepository,
)

_ACCOUNT_ID = "000000000000"
_REGION = "us-east-1"


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _topic_arn(name: str) -> str:
    return f"arn:aws:sns:{_REGION}:{_ACCOUNT_ID}:{name}"


def _subscription_arn(topic_arn: str, suffix: str) -> str:
    return f"{topic_arn}:{suffix}"


class SnsService:
    def __init__(
        self,
        topic_repo: SnsTopicRepository,
        subscription_repo: SnsSubscriptionRepository,
        message_repo: SnsMessageRepository,
        telemetry: TelemetryEngine,
    ):
        self._topic_repo = topic_repo
        self._subscription_repo = subscription_repo
        self._message_repo = message_repo
        self._telemetry = telemetry

    # -------------------------------------------------------------------
    # Topics
    # -------------------------------------------------------------------

    async def create_topic(self, name: str) -> str:
        """Idempotent – returns the ARN of an existing topic if it already exists."""
        existing = await self._topic_repo.get_by_name(name)
        if existing:
            return existing.arn

        arn = _topic_arn(name)
        topic = SnsTopic(id=None, name=name, arn=arn, created_at=_now())
        await self._topic_repo.save(topic)
        await self._telemetry.emit("aws", "sns", "create_topic", {"name": name, "arn": arn})
        return arn

    async def list_topics(self) -> list[str]:
        """Return all topic ARNs."""
        topics = await self._topic_repo.list_all()
        return [t.arn for t in topics]

    # -------------------------------------------------------------------
    # Subscriptions
    # -------------------------------------------------------------------

    async def subscribe(self, topic_arn: str, protocol: str, endpoint: str) -> str:
        """Subscribe an endpoint to a topic. Returns the SubscriptionArn."""
        topic = await self._topic_repo.get(topic_arn)
        if topic is None:
            raise NotFoundError(f"Topic not found: {topic_arn}")

        # Check for duplicate (same topic + protocol + endpoint)
        existing = await self._subscription_repo.list_by_topic(topic_arn)
        for sub in existing:
            if sub.protocol == protocol and sub.endpoint == endpoint:
                return sub.subscription_arn

        sub_arn = _subscription_arn(topic_arn, str(uuid.uuid4()))
        subscription = SnsSubscription(
            id=None,
            subscription_arn=sub_arn,
            topic_arn=topic_arn,
            protocol=protocol,
            endpoint=endpoint,
            created_at=_now(),
        )
        await self._subscription_repo.save(subscription)
        await self._telemetry.emit(
            "aws", "sns", "subscribe",
            {"topic_arn": topic_arn, "protocol": protocol, "endpoint": endpoint},
        )
        return sub_arn

    # -------------------------------------------------------------------
    # Publish
    # -------------------------------------------------------------------

    async def publish(
        self,
        topic_arn: str,
        message: str,
        subject: str | None = None,
    ) -> str:
        """Publish a message to a topic. Returns the MessageId."""
        topic = await self._topic_repo.get(topic_arn)
        if topic is None:
            raise NotFoundError(f"Topic not found: {topic_arn}")

        message_id = str(uuid.uuid4())
        msg = SnsMessage(
            id=None,
            message_id=message_id,
            topic_arn=topic_arn,
            message=message,
            subject=subject,
            created_at=_now(),
        )
        await self._message_repo.save(msg)
        await self._telemetry.emit(
            "aws", "sns", "publish",
            {"topic_arn": topic_arn, "message_id": message_id},
        )
        return message_id

File: rust/cloudtwin-lite/src/aws/sns/service.rs
//! SNS business logic.
//!
//! This service persists topics, subscriptions, and published messages while
//! staying agnostic of the AWS Query HTTP envelope.

use anyhow::{bail, Result};
use chrono::Utc;
use uuid::Uuid;

use super::models::{SnsSubscription, SnsTopic};
use crate::db::Database;

const ACCOUNT_ID: &str = "000000000000";
const REGION: &str = "us-east-1";

/// Encapsulates SNS topic, subscription, and publish operations.
pub struct SnsService {
    db: Database,
}

impl SnsService {
    pub fn new(db: Database) -> Self {
        SnsService { db }
    }

    fn topic_arn(name: &str) -> String {
        format!("arn:aws:sns:{REGION}:{ACCOUNT_ID}:{name}")
    }
    #[allow(dead_code)]
    fn sub_arn(name: &str) -> String {
        format!(
            "arn:aws:sns:{REGION}:{ACCOUNT_ID}:{name}:{}",
            Uuid::new_v4().simple()
        )
    }

    pub async fn create_topic(&self, name: &str) -> Result<String> {
        let arn = Self::topic_arn(name);
        let n = name.to_string();
        let a = arn.clone();
        let now = Utc::now().to_rfc3339();
        self.db
            .conn
            .call(move |conn| {
                conn.execute(
                    "INSERT INTO sns_topics (arn, name, created_at)
                 VALUES (?1, ?2, ?3)
                 ON CONFLICT(name) DO NOTHING",
                    rusqlite::params![a, n, now],
                )?;
                Ok(())
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(arn)
    }

    pub async fn list_topics(&self) -> Result<Vec<SnsTopic>> {
        self.db
            .conn
            .call(|conn| {
                let mut stmt =
                    conn.prepare("SELECT arn, name, created_at FROM sns_topics ORDER BY name")?;
                let rows = stmt
                    .query_map([], |r| {
                        Ok(SnsTopic {
                            arn: r.get(0)?,
                            name: r.get(1)?,
                            created_at: r.get(2)?,
                        })
                    })?
                    .collect::<rusqlite::Result<Vec<_>>>()?;
                Ok(rows)
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn delete_topic(&self, arn: &str) -> Result<()> {
        let a = arn.to_string();
        self.db
            .conn
            .call(move |conn| {
                conn.execute(
                    "DELETE FROM sns_topics WHERE arn = ?1",
                    rusqlite::params![a],
                )?;
                Ok(())
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn subscribe(
        &self,
        topic_arn: &str,
        protocol: &str,
        endpoint: &str,
    ) -> Result<String> {
        // Validate topic exists
        let ta = topic_arn.to_string();
        let exists: bool = self
            .db
            .conn
            .call(move |conn| {
                let n: i64 = conn.query_row(
                    "SELECT COUNT(*) FROM sns_topics WHERE arn = ?1",
                    rusqlite::params![ta],
                    |r| r.get(0),
                )?;
                Ok(n > 0)
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))?;
        if !exists {
            bail!("NotFound: topic {topic_arn}");
        }

        // Derive sub name from endpoint
        let sub_name = topic_arn.rsplit(':').next().unwrap_or("topic");
        let sub_arn = format!(
            "arn:aws:sns:{REGION}:{ACCOUNT_ID}:{sub_name}:{}",
            Uuid::new_v4().simple()
        );
        let ta2 = topic_arn.to_string();
        let pr = protocol.to_string();
        let ep = endpoint.to_string();
        let sa = sub_arn.clone();
        let now = Utc::now().to_rfc3339();
        self.db
            .conn
            .call(move |conn| {
                conn.execute(
                    "INSERT INTO sns_subscriptions (arn, topic_arn, protocol, endpoint, created_at)
                 VALUES (?1, ?2, ?3, ?4, ?5)
                 ON CONFLICT(arn) DO NOTHING",
                    rusqlite::params![sa, ta2, pr, ep, now],
                )?;
                Ok(())
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(sub_arn)
    }

    pub async fn list_subscriptions_by_topic(
        &self,
        topic_arn: &str,
    ) -> Result<Vec<SnsSubscription>> {
        let ta = topic_arn.to_string();
        self.db
            .conn
            .call(move |conn| {
                let mut stmt = conn.prepare(
                    "SELECT arn, topic_arn, protocol, endpoint, created_at
                 FROM sns_subscriptions WHERE topic_arn = ?1 ORDER BY arn",
                )?;
                let rows = stmt
                    .query_map(rusqlite::params![ta], |r| {
                        Ok(SnsSubscription {
                            arn: r.get(0)?,
                            topic_arn: r.get(1)?,
                            protocol: r.get(2)?,
                            endpoint: r.get(3)?,
                            created_at: r.get(4)?,
                        })
                    })?
                    .collect::<rusqlite::Result<Vec<_>>>()?;
                Ok(rows)
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn publish(
        &self,
        topic_arn: &str,
        message: &str,
        subject: Option<&str>,
    ) -> Result<String> {
        let message_id = Uuid::new_v4().to_string();
        let ta = topic_arn.to_string();
        let mid = message_id.clone();
        let msg = message.to_string();
        let subj = subject.map(|s| s.to_string());
        let now = Utc::now().to_rfc3339();
        self.db
            .conn
            .call(move |conn| {
                conn.execute(
                    "INSERT INTO sns_messages (message_id, topic_arn, subject, body, created_at)
                 VALUES (?1, ?2, ?3, ?4, ?5)",
                    rusqlite::params![mid, ta, subj, msg, now],
                )?;
                Ok(())
            })
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(message_id)
    }
}
